File: ruby/http_benchmark.rb
require 'net/http'

raise ArgumentError, 'No URL provided' unless ARGV[0]

URL = URI(ARGV[0]).freeze
results = []

trap "SIGINT" do
  print_results(results)
  exit 130
end

def print_results(results)
  if results.size == 0
    puts 'Nothing to print...'
    return
  end

  puts "Bechmarking HTTP GET to #{URL}"
  puts "avg #{results.sum(0.0) / results.size}"
  puts "min #{results.min}"
  puts "max #{results.max}"
  puts "#{results.count} requests total"
end

while true do
  time = Time.now
  response = Net::HTTP.get_response(URL)
  response.body
  diff = Time.now - time
  puts "Fetched in #{diff} seconds"
  results << diff
end

File: go/http_benchmark.go
package main

import (
	"fmt"
	"net/http"
	"io"
	"os"
	"os/signal"
	"syscall"
	"time"
)

func doRequest(url string) (time.Duration) {
	start := time.Now()

	resp, err := http.Get(url)
	if err != nil {
		fmt.Println("Error while fetching")
	}
	defer resp.Body.Close()
	_, readErr := io.ReadAll(resp.Body)
	if readErr != nil {
		fmt.Println(readErr)
	}

	end := time.Now()
	elapsed := end.Sub(start)

	s := fmt.Sprintf("Fetched in %v", elapsed.Round(time.Millisecond))
	fmt.Println(s)

	return elapsed
}

func findMinAvgMax(results []int) (min int, avg float64, max int) {
	sum := 0
	min = results[0]
	max = results[0]
	for _, value := range results {
		sum += value
		if value < min {
			min = value
		}
		if value > max {
			max = value
		}
	}
	avg = float64(sum) / float64(len(results))

	return min, avg, max
}

func printResults(results []int) {
	min, avg, max := findMinAvgMax(results)
	s := fmt.Sprintf("Min: %v ms\nAvg: %v ms\nMax: %v ms\nTotal requests: %d", min, avg, max, len(results))
	fmt.Println(s)
}

func toMs(duration time.Duration) (int) {
	return int(duration / time.Millisecond)
}

func main() {
	url := os.Args[1]
	c := make(chan os.Signal)
	var results []int
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
			<-c
			printResults(results)
			os.Exit(1)
	}()

	for {
		results = append(results, toMs(doRequest(url)))
	}
}
